raft/raft.cpp

```diff
@@ -407,7 +416,85 @@ raft::create_joint_quorum_by_removing_peer(const bzn::message& last_quorum_messa
 }
 
 
+void
+raft::handle_add_peer(std::shared_ptr<bzn::session_base> session, const bzn::message &peer)
+{
+    if (this->get_state() != bzn::raft_state::leader)
+    {
+        bzn::message response;
+        response["error"] = ERROR_ADD_PEER_MUST_BE_SENT_TO_LEADER;
+        session->send_message(std::make_shared<bzn::message>(response), true);
+        return;
+    }
+
+    bzn::log_entry last_quorum_entry = this->raft_log->last_quorum_entry();
+    if (last_quorum_entry.entry_type == bzn::log_entry_type::joint_quorum)
+    {
+        bzn::message response;
+        response["error"] = MSG_ERROR_CURRENT_QUORUM_IS_JOINT;
+        session->send_message(std::make_shared<bzn::message>(response), true);
+        return;
+    }
+
+    const auto &same_peer = std::find_if(last_quorum_entry.msg["msg"]["peers"].begin(),
+                                         last_quorum_entry.msg["msg"]["peers"].end(),
+                                         [&](const auto &p)
+                                         {
+                                             return p["uuid"].asString() == peer["uuid"].asString();
+                                         });
+    if (same_peer != last_quorum_entry.msg["msg"]["peers"].end())
+    {
+        bzn::message response;
+        response["error"] = ERROR_PEER_ALREADY_EXISTS;
+        session->send_message(std::make_shared<bzn::message>(response), true);
+        return;
+    }
 
+    this->peer_match_index[peer["uuid"].asString()] = 1;
```

Comment:
should we check that uuid is there otherwise we'll get an empty string back?

Reply:
I have added a catch that will check if "uuid" does not exist, or the value is empty, then add peer will fail with a message.

Also created unit tests to exercise the bad add and remove paths... 

But, for UUID only. I will sneak in the tests for all the other parameters in future refactorings.